From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 21:00:08 -0700
Subject: [PATCH] mshtml: element events, ready states, anchor and area address
 parts

onselect/onerror/onabort/onreadystatechange/onunload/onbeforeunload/onscroll/
onreset/onchange/onblur/onfocus of the elements that lacked them,
readyState of input, img, link and object, the anchor and area setters of
host, hostname, port, pathname, protocol, search and hash (through the
address), the area interface's missing hostname slot in nsiface.idl, and
Clone for the form and select enumerators.

diff --git a/dlls/mshtml/htmlanchor.c b/dlls/mshtml/htmlanchor.c
index df770ce..f2a085a 100644
--- a/dlls/mshtml/htmlanchor.c
+++ b/dlls/mshtml/htmlanchor.c
@@ -197,6 +197,126 @@ static IUri *get_anchor_uri(HTMLAnchorElement *anchor)
     return uri;
 }
 
+/* An address with one of its parts replaced (the setters of the host, port, ... of an
+ * anchor or an area), as the new address. */
+HRESULT url_with_part(const WCHAR *href, enum sg_url_part part, const WCHAR *value, BSTR *ret)
+{
+    BSTR scheme = NULL, user = NULL, host = NULL, path = NULL, query = NULL, fragment = NULL, abs = NULL;
+    DWORD port = 0;
+    BOOL has_port = FALSE;
+    WCHAR *out, *colon, *copy = NULL;
+    size_t len;
+    IUri *uri;
+    HRESULT hres;
+
+    *ret = NULL;
+    if(!href || !*href)
+        return E_FAIL;
+    hres = create_uri(href, 0, &uri);
+    if(FAILED(hres))
+        return hres;
+
+    IUri_GetSchemeName(uri, &scheme);
+    IUri_GetUserInfo(uri, &user);
+    IUri_GetHost(uri, &host);
+    IUri_GetPath(uri, &path);
+    IUri_GetQuery(uri, &query);
+    IUri_GetFragment(uri, &fragment);
+    IUri_GetAbsoluteUri(uri, &abs);
+    has_port = IUri_GetPort(uri, &port) == S_OK;
+    IUri_Release(uri);
+
+    if(!value) value = L"";
+    switch(part) {
+    case SG_URLPART_PORT:
+#ifdef SG_MUTANT_MSHTML_URL_NOPORT
+        break;
+#endif
+        has_port = *value != 0;
+        port = wcstoul(value, NULL, 10);
+        break;
+    case SG_URLPART_HOSTNAME:
+        SysFreeString(host);
+        host = SysAllocString(value);
+        break;
+    case SG_URLPART_HOST:
+        copy = wcsdup(value);
+        if(copy && (colon = wcsrchr(copy, ':'))) {
+            *colon = 0;
+            has_port = colon[1] != 0;
+            port = wcstoul(colon + 1, NULL, 10);
+        }
+        SysFreeString(host);
+        host = SysAllocString(copy ? copy : L"");
+        free(copy);
+        break;
+    case SG_URLPART_PATHNAME:
+        SysFreeString(path);
+        path = SysAllocString(*value == '/' ? value : L"");
+        if(*value && *value != '/') {
+            SysFreeString(path);
+            path = SysAllocStringLen(NULL, wcslen(value) + 1);
+            if(path) swprintf(path, wcslen(value) + 2, L"/%s", value);
+        }
+        break;
+    case SG_URLPART_PROTOCOL:
+        SysFreeString(scheme);
+        scheme = SysAllocString(value);
+        if(scheme && SysStringLen(scheme) && scheme[SysStringLen(scheme) - 1] == ':')
+            scheme[SysStringLen(scheme) - 1] = 0;
+        break;
+    case SG_URLPART_SEARCH:
+        SysFreeString(query);
+        query = SysAllocString(!*value || *value == '?' ? value : L"");
+        if(*value && *value != '?') {
+            SysFreeString(query);
+            query = SysAllocStringLen(NULL, wcslen(value) + 1);
+            if(query) swprintf(query, wcslen(value) + 2, L"?%s", value);
+        }
+        break;
+    case SG_URLPART_HASH:
+        SysFreeString(fragment);
+        fragment = SysAllocString(!*value || *value == '#' ? value : L"");
+        if(*value && *value != '#') {
+            SysFreeString(fragment);
+            fragment = SysAllocStringLen(NULL, wcslen(value) + 1);
+            if(fragment) swprintf(fragment, wcslen(value) + 2, L"#%s", value);
+        }
+        break;
+    }
+
+    /* scheme://[userinfo@]host[:port]path?query#fragment; the query and the fragment keep their marks */
+    len = SysStringLen(scheme) + SysStringLen(user) + SysStringLen(host) + SysStringLen(path)
+        + SysStringLen(query) + SysStringLen(fragment) + 32;
+    if((out = malloc(len * sizeof(WCHAR)))) {
+        const BOOL hierarchical = abs && wcsstr(abs, L"://");
+        const WCHAR *sep = hierarchical || (host && *host) ? L"//" : L"";
+        int n = swprintf(out, len, L"%s:%s", scheme ? scheme : L"", sep);
+
+        if(user && *user)
+            n += swprintf(out + n, len - n, L"%s@", user);
+        if(host)
+            n += swprintf(out + n, len - n, L"%s", host);
+        if(has_port && host && *host)
+            n += swprintf(out + n, len - n, L":%lu", port);
+        if(path)
+            n += swprintf(out + n, len - n, L"%s", path);
+        if(query)
+            n += swprintf(out + n, len - n, L"%s", query);
+        if(fragment)
+            swprintf(out + n, len - n, L"%s", fragment);
+        *ret = SysAllocString(out);
+        free(out);
+        hres = *ret ? S_OK : E_OUTOFMEMORY;
+    }else {
+        hres = E_OUTOFMEMORY;
+    }
+
+    SysFreeString(scheme); SysFreeString(user); SysFreeString(host); SysFreeString(path);
+    SysFreeString(query); SysFreeString(fragment); SysFreeString(abs);
+    return hres;
+}
+
 static inline HTMLAnchorElement *impl_from_IHTMLAnchorElement(IHTMLAnchorElement *iface)
 {
     return CONTAINING_RECORD(iface, HTMLAnchorElement, IHTMLAnchorElement_iface);
@@ -391,11 +511,38 @@ static HRESULT WINAPI HTMLAnchorElement_get_name(IHTMLAnchorElement *iface, BSTR
     return return_nsstr(nsres, &name_str, p);
 }
 
+static HRESULT anchor_set_part(IHTMLAnchorElement *iface, HTMLAnchorElement *This, enum sg_url_part part, BSTR v)
+{
+    nsAString href_str;
+    BSTR href = NULL;
+    nsresult nsres;
+    HRESULT hres;
+
+    nsAString_Init(&href_str, NULL);
+    nsres = nsIDOMHTMLAnchorElement_GetHref(This->nsanchor, &href_str);
+    if(NS_SUCCEEDED(nsres)) {
+        const PRUnichar *str;
+        nsAString_GetData(&href_str, &str);
+        hres = url_with_part(str, part, v, &href);
+    }else {
+        hres = E_FAIL;
+    }
+    nsAString_Finish(&href_str);
+    if(FAILED(hres))
+        return hres;
+
+    hres = IHTMLAnchorElement_put_href(iface, href);
+    SysFreeString(href);
+    return hres;
+}
+
 static HRESULT WINAPI HTMLAnchorElement_put_host(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_HOST, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_host(IHTMLAnchorElement *iface, BSTR *p)
@@ -415,8 +562,10 @@ static HRESULT WINAPI HTMLAnchorElement_get_host(IHTMLAnchorElement *iface, BSTR
 static HRESULT WINAPI HTMLAnchorElement_put_hostname(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_HOSTNAME, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_hostname(IHTMLAnchorElement *iface, BSTR *p)
@@ -435,8 +584,10 @@ static HRESULT WINAPI HTMLAnchorElement_get_hostname(IHTMLAnchorElement *iface,
 static HRESULT WINAPI HTMLAnchorElement_put_pathname(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_PATHNAME, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_pathname(IHTMLAnchorElement *iface, BSTR *p)
@@ -456,8 +607,10 @@ static HRESULT WINAPI HTMLAnchorElement_get_pathname(IHTMLAnchorElement *iface,
 static HRESULT WINAPI HTMLAnchorElement_put_port(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_PORT, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_port(IHTMLAnchorElement *iface, BSTR *p)
@@ -501,8 +654,10 @@ static HRESULT WINAPI HTMLAnchorElement_get_port(IHTMLAnchorElement *iface, BSTR
 static HRESULT WINAPI HTMLAnchorElement_put_protocol(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_PROTOCOL, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_protocol(IHTMLAnchorElement *iface, BSTR *p)
@@ -579,8 +734,10 @@ static HRESULT WINAPI HTMLAnchorElement_get_search(IHTMLAnchorElement *iface, BS
 static HRESULT WINAPI HTMLAnchorElement_put_hash(IHTMLAnchorElement *iface, BSTR v)
 {
     HTMLAnchorElement *This = impl_from_IHTMLAnchorElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    return anchor_set_part(iface, This, SG_URLPART_HASH, v);
 }
 
 static HRESULT WINAPI HTMLAnchorElement_get_hash(IHTMLAnchorElement *iface, BSTR *p)
diff --git a/dlls/mshtml/htmlarea.c b/dlls/mshtml/htmlarea.c
index 66f1a4c..ca21f7d 100644
--- a/dlls/mshtml/htmlarea.c
+++ b/dlls/mshtml/htmlarea.c
@@ -189,127 +189,268 @@ static HRESULT WINAPI HTMLAreaElement_get_noHref(IHTMLAreaElement *iface, VARIAN
 static HRESULT WINAPI HTMLAreaElement_put_host(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_HOST, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_host(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetHost(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_hostname(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_HOSTNAME, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_hostname(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetHostname(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_pathname(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_PATHNAME, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_pathname(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetPathname(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_port(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_PORT, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_port(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetPort(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_protocol(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_PROTOCOL, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_protocol(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetProtocol(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_search(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_SEARCH, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_search(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetSearch(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_hash(IHTMLAreaElement *iface, BSTR v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    BSTR href, new_href;
+    HRESULT hres;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_w(v));
+
+    hres = IHTMLAreaElement_get_href(iface, &href);
+    if(FAILED(hres))
+        return hres;
+    hres = url_with_part(href, SG_URLPART_HASH, v, &new_href);
+    SysFreeString(href);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLAreaElement_put_href(iface, new_href);
+    SysFreeString(new_href);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_hash(IHTMLAreaElement *iface, BSTR *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    nsAString nsstr;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsAString_Init(&nsstr, NULL);
+    nsres = nsIDOMHTMLAreaElement_GetHash(This->nsarea, &nsstr);
+    return return_nsstr(nsres, &nsstr, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_onblur(IHTMLAreaElement *iface, VARIANT v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, &v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_BLUR, &v);
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_onblur(IHTMLAreaElement *iface, VARIANT *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_BLUR, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_onfocus(IHTMLAreaElement *iface, VARIANT v)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, &v);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_FOCUS, &v);
 }
 
 static HRESULT WINAPI HTMLAreaElement_get_onfocus(IHTMLAreaElement *iface, VARIANT *p)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_FOCUS, p);
 }
 
 static HRESULT WINAPI HTMLAreaElement_put_tabIndex(IHTMLAreaElement *iface, short v)
@@ -334,15 +475,19 @@ static HRESULT WINAPI HTMLAreaElement_get_tabIndex(IHTMLAreaElement *iface, shor
 static HRESULT WINAPI HTMLAreaElement_focus(IHTMLAreaElement *iface)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)\n", This);
+
+    return IHTMLElement2_focus(&This->element.IHTMLElement2_iface);
 }
 
 static HRESULT WINAPI HTMLAreaElement_blur(IHTMLAreaElement *iface)
 {
     HTMLAreaElement *This = impl_from_IHTMLAreaElement(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)\n", This);
+
+    return IHTMLElement2_blur(&This->element.IHTMLElement2_iface);
 }
 
 static const IHTMLAreaElementVtbl HTMLAreaElementVtbl = {
diff --git a/dlls/mshtml/htmlbody.c b/dlls/mshtml/htmlbody.c
index b45df67..c8c2c7e 100644
--- a/dlls/mshtml/htmlbody.c
+++ b/dlls/mshtml/htmlbody.c
@@ -539,15 +539,19 @@ static HRESULT WINAPI HTMLBodyElement_get_onload(IHTMLBodyElement *iface, VARIAN
 static HRESULT WINAPI HTMLBodyElement_put_onunload(IHTMLBodyElement *iface, VARIANT v)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_UNLOAD, &v);
 }
 
 static HRESULT WINAPI HTMLBodyElement_get_onunload(IHTMLBodyElement *iface, VARIANT *p)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_UNLOAD, p);
 }
 
 static HRESULT WINAPI HTMLBodyElement_put_scroll(IHTMLBodyElement *iface, BSTR v)
@@ -613,29 +617,37 @@ static HRESULT WINAPI HTMLBodyElement_get_scroll(IHTMLBodyElement *iface, BSTR *
 static HRESULT WINAPI HTMLBodyElement_put_onselect(IHTMLBodyElement *iface, VARIANT v)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_SELECT, &v);
 }
 
 static HRESULT WINAPI HTMLBodyElement_get_onselect(IHTMLBodyElement *iface, VARIANT *p)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_SELECT, p);
 }
 
 static HRESULT WINAPI HTMLBodyElement_put_onbeforeunload(IHTMLBodyElement *iface, VARIANT v)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_BEFOREUNLOAD, &v);
 }
 
 static HRESULT WINAPI HTMLBodyElement_get_onbeforeunload(IHTMLBodyElement *iface, VARIANT *p)
 {
     HTMLBodyElement *This = impl_from_IHTMLBodyElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_BEFOREUNLOAD, p);
 }
 
 static HRESULT WINAPI HTMLBodyElement_createTextRange(IHTMLBodyElement *iface, IHTMLTxtRange **range)
@@ -785,15 +797,19 @@ static HRESULT WINAPI HTMLTextContainer_get_scrollLeft(IHTMLTextContainer *iface
 static HRESULT WINAPI HTMLTextContainer_put_onscroll(IHTMLTextContainer *iface, VARIANT v)
 {
     HTMLBodyElement *This = impl_from_IHTMLTextContainer(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_SCROLL, &v);
 }
 
 static HRESULT WINAPI HTMLTextContainer_get_onscroll(IHTMLTextContainer *iface, VARIANT *p)
 {
     HTMLBodyElement *This = impl_from_IHTMLTextContainer(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_SCROLL, p);
 }
 
 static const IHTMLTextContainerVtbl HTMLTextContainerVtbl = {
diff --git a/dlls/mshtml/htmlform.c b/dlls/mshtml/htmlform.c
index 5fbe917..da14cd2 100644
--- a/dlls/mshtml/htmlform.c
+++ b/dlls/mshtml/htmlform.c
@@ -232,8 +232,21 @@ static HRESULT WINAPI HTMLFormElementEnum_Reset(IEnumVARIANT *iface)
 static HRESULT WINAPI HTMLFormElementEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
 {
     HTMLFormElementEnum *This = impl_from_IEnumVARIANT(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
-    return E_NOTIMPL;
+    HTMLFormElementEnum *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumVARIANT_iface.lpVtbl = This->IEnumVARIANT_iface.lpVtbl;
+    ret->ref = 1;
+    ret->iter = This->iter;
+    IHTMLFormElement_AddRef(&This->elem->IHTMLFormElement_iface);
+    ret->elem = This->elem;
+
+    *ppEnum = &ret->IEnumVARIANT_iface;
+    return S_OK;
 }
 
 static const IEnumVARIANTVtbl HTMLFormElementEnumVtbl = {
@@ -497,15 +510,19 @@ static HRESULT WINAPI HTMLFormElement_get_onsubmit(IHTMLFormElement *iface, VARI
 static HRESULT WINAPI HTMLFormElement_put_onreset(IHTMLFormElement *iface, VARIANT v)
 {
     HTMLFormElement *This = impl_from_IHTMLFormElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_RESET, &v);
 }
 
 static HRESULT WINAPI HTMLFormElement_get_onreset(IHTMLFormElement *iface, VARIANT *p)
 {
     HTMLFormElement *This = impl_from_IHTMLFormElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_RESET, p);
 }
 
 static HRESULT WINAPI HTMLFormElement_submit(IHTMLFormElement *iface)
diff --git a/dlls/mshtml/htmlframe.c b/dlls/mshtml/htmlframe.c
index f6e20a1..34017e0 100644
--- a/dlls/mshtml/htmlframe.c
+++ b/dlls/mshtml/htmlframe.c
@@ -600,15 +600,19 @@ static HRESULT WINAPI HTMLFrameBase2_get_onload(IHTMLFrameBase2 *iface, VARIANT
 static HRESULT WINAPI HTMLFrameBase2_put_onreadystatechange(IHTMLFrameBase2 *iface, VARIANT v)
 {
     HTMLFrameBase *This = impl_from_IHTMLFrameBase2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_READYSTATECHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLFrameBase2_get_onreadystatechange(IHTMLFrameBase2 *iface, VARIANT *p)
 {
     HTMLFrameBase *This = impl_from_IHTMLFrameBase2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_READYSTATECHANGE, p);
 }
 
 static HRESULT WINAPI HTMLFrameBase2_get_readyState(IHTMLFrameBase2 *iface, BSTR *p)
diff --git a/dlls/mshtml/htmlimg.c b/dlls/mshtml/htmlimg.c
index 59dd83b..93dc3c0 100644
--- a/dlls/mshtml/htmlimg.c
+++ b/dlls/mshtml/htmlimg.c
@@ -354,8 +354,28 @@ static HRESULT WINAPI HTMLImgElement_get_dynsrc(IHTMLImgElement *iface, BSTR *p)
 static HRESULT WINAPI HTMLImgElement_get_readyState(IHTMLImgElement *iface, BSTR *p)
 {
     HTMLImg *This = impl_from_IHTMLImgElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    cpp_bool complete = FALSE;
+    nsAString src;
+    const WCHAR *state;
+    nsresult nsres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    nsres = nsIDOMHTMLImageElement_GetComplete(This->nsimg, &complete);
+    if(NS_FAILED(nsres))
+        return E_FAIL;
+    if(complete) {
+        state = L"complete";
+    }else {
+        const WCHAR *str;
+        nsAString_Init(&src, NULL);
+        nsIDOMHTMLImageElement_GetSrc(This->nsimg, &src);
+        nsAString_GetData(&src, &str);
+        state = *str ? L"loading" : L"uninitialized";
+        nsAString_Finish(&src);
+    }
+    *p = SysAllocString(state);
+    return *p ? S_OK : E_OUTOFMEMORY;
 }
 
 static HRESULT WINAPI HTMLImgElement_get_complete(IHTMLImgElement *iface, VARIANT_BOOL *p)
diff --git a/dlls/mshtml/htmlinput.c b/dlls/mshtml/htmlinput.c
index 8a25766..3c77ba2 100644
--- a/dlls/mshtml/htmlinput.c
+++ b/dlls/mshtml/htmlinput.c
@@ -314,15 +314,19 @@ static HRESULT WINAPI HTMLInputElement_get_onchange(IHTMLInputElement *iface, VA
 static HRESULT WINAPI HTMLInputElement_put_onselect(IHTMLInputElement *iface, VARIANT v)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_SELECT, &v);
 }
 
 static HRESULT WINAPI HTMLInputElement_get_onselect(IHTMLInputElement *iface, VARIANT *p)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_SELECT, p);
 }
 
 static HRESULT WINAPI HTMLInputElement_put_defaultValue(IHTMLInputElement *iface, BSTR v)
@@ -657,8 +661,22 @@ static HRESULT WINAPI HTMLInputElement_get_dynsrc(IHTMLInputElement *iface, BSTR
 static HRESULT WINAPI HTMLInputElement_get_readyState(IHTMLInputElement *iface, BSTR *p)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLElement2 *elem2;
+    VARIANT v;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    hres = IHTMLElement_QueryInterface(&This->element.IHTMLElement_iface, &IID_IHTMLElement2, (void**)&elem2);
+    if(FAILED(hres))
+        return hres;
+    V_VT(&v) = VT_EMPTY;
+    hres = IHTMLElement2_get_readyState(elem2, &v);
+    IHTMLElement2_Release(elem2);
+    if(FAILED(hres))
+        return hres;
+    *p = V_BSTR(&v);
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLInputElement_get_complete(IHTMLInputElement *iface, VARIANT_BOOL *p)
@@ -736,29 +754,37 @@ static HRESULT WINAPI HTMLInputElement_get_onload(IHTMLInputElement *iface, VARI
 static HRESULT WINAPI HTMLInputElement_put_onerror(IHTMLInputElement *iface, VARIANT v)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_ERROR, &v);
 }
 
 static HRESULT WINAPI HTMLInputElement_get_onerror(IHTMLInputElement *iface, VARIANT *p)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_ERROR, p);
 }
 
 static HRESULT WINAPI HTMLInputElement_put_onabort(IHTMLInputElement *iface, VARIANT v)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_ABORT, &v);
 }
 
 static HRESULT WINAPI HTMLInputElement_get_onabort(IHTMLInputElement *iface, VARIANT *p)
 {
     HTMLInputElement *This = impl_from_IHTMLInputElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_ABORT, p);
 }
 
 static HRESULT WINAPI HTMLInputElement_put_width(IHTMLInputElement *iface, LONG v)
diff --git a/dlls/mshtml/htmllink.c b/dlls/mshtml/htmllink.c
index c9e35d6..4f3946d 100644
--- a/dlls/mshtml/htmllink.c
+++ b/dlls/mshtml/htmllink.c
@@ -166,22 +166,40 @@ static HRESULT WINAPI HTMLLinkElement_get_type(IHTMLLinkElement *iface, BSTR *p)
 static HRESULT WINAPI HTMLLinkElement_get_readyState(IHTMLLinkElement *iface, BSTR *p)
 {
     HTMLLinkElement *This = impl_from_IHTMLLinkElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLElement2 *elem2;
+    VARIANT v;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    hres = IHTMLElement_QueryInterface(&This->element.IHTMLElement_iface, &IID_IHTMLElement2, (void**)&elem2);
+    if(FAILED(hres))
+        return hres;
+    V_VT(&v) = VT_EMPTY;
+    hres = IHTMLElement2_get_readyState(elem2, &v);
+    IHTMLElement2_Release(elem2);
+    if(FAILED(hres))
+        return hres;
+    *p = V_BSTR(&v);
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLLinkElement_put_onreadystatechange(IHTMLLinkElement *iface, VARIANT v)
 {
     HTMLLinkElement *This = impl_from_IHTMLLinkElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_READYSTATECHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLLinkElement_get_onreadystatechange(IHTMLLinkElement *iface, VARIANT *p)
 {
     HTMLLinkElement *This = impl_from_IHTMLLinkElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_READYSTATECHANGE, p);
 }
 
 static HRESULT WINAPI HTMLLinkElement_put_onload(IHTMLLinkElement *iface, VARIANT v)
@@ -205,15 +223,19 @@ static HRESULT WINAPI HTMLLinkElement_get_onload(IHTMLLinkElement *iface, VARIAN
 static HRESULT WINAPI HTMLLinkElement_put_onerror(IHTMLLinkElement *iface, VARIANT v)
 {
     HTMLLinkElement *This = impl_from_IHTMLLinkElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_ERROR, &v);
 }
 
 static HRESULT WINAPI HTMLLinkElement_get_onerror(IHTMLLinkElement *iface, VARIANT *p)
 {
     HTMLLinkElement *This = impl_from_IHTMLLinkElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_ERROR, p);
 }
 
 static HRESULT WINAPI HTMLLinkElement_get_styleSheet(IHTMLLinkElement *iface, IHTMLStyleSheet **p)
diff --git a/dlls/mshtml/htmlobject.c b/dlls/mshtml/htmlobject.c
index 91ff340..0eadb91 100644
--- a/dlls/mshtml/htmlobject.c
+++ b/dlls/mshtml/htmlobject.c
@@ -351,36 +351,53 @@ static HRESULT WINAPI HTMLObjectElement_get_height(IHTMLObjectElement *iface, VA
 static HRESULT WINAPI HTMLObjectElement_get_readyState(IHTMLObjectElement *iface, LONG *p)
 {
     HTMLObjectElement *This = impl_from_IHTMLObjectElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLElement2 *elem2;
+    HRESULT hres;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    hres = IHTMLElement_QueryInterface(&This->plugin_container.element.IHTMLElement_iface, &IID_IHTMLElement2, (void**)&elem2);
+    if(FAILED(hres))
+        return hres;
+    hres = IHTMLElement2_get_readyStateValue(elem2, p);
+    IHTMLElement2_Release(elem2);
+    return hres;
 }
 
 static HRESULT WINAPI HTMLObjectElement_put_onreadystatechange(IHTMLObjectElement *iface, VARIANT v)
 {
     HTMLObjectElement *This = impl_from_IHTMLObjectElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->plugin_container.element.node, EVENTID_READYSTATECHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLObjectElement_get_onreadystatechange(IHTMLObjectElement *iface, VARIANT *p)
 {
     HTMLObjectElement *This = impl_from_IHTMLObjectElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->plugin_container.element.node, EVENTID_READYSTATECHANGE, p);
 }
 
 static HRESULT WINAPI HTMLObjectElement_put_onerror(IHTMLObjectElement *iface, VARIANT v)
 {
     HTMLObjectElement *This = impl_from_IHTMLObjectElement(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->plugin_container.element.node, EVENTID_ERROR, &v);
 }
 
 static HRESULT WINAPI HTMLObjectElement_get_onerror(IHTMLObjectElement *iface, VARIANT *p)
 {
     HTMLObjectElement *This = impl_from_IHTMLObjectElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->plugin_container.element.node, EVENTID_ERROR, p);
 }
 
 static HRESULT WINAPI HTMLObjectElement_put_altHtml(IHTMLObjectElement *iface, BSTR v)
@@ -636,8 +653,11 @@ static inline HTMLObjectElement *impl_from_HTMLDOMNode(HTMLDOMNode *iface)
 static HRESULT HTMLObjectElement_get_readystate(HTMLDOMNode *iface, BSTR *p)
 {
     HTMLObjectElement *This = impl_from_HTMLDOMNode(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    *p = SysAllocString(L"complete");
+    return *p ? S_OK : E_OUTOFMEMORY;
 }
 
 static inline HTMLObjectElement *impl_from_DispatchEx(DispatchEx *iface)
diff --git a/dlls/mshtml/htmlselect.c b/dlls/mshtml/htmlselect.c
index 2d02589..60fd326 100644
--- a/dlls/mshtml/htmlselect.c
+++ b/dlls/mshtml/htmlselect.c
@@ -722,8 +722,21 @@ static HRESULT WINAPI HTMLSelectElementEnum_Reset(IEnumVARIANT *iface)
 static HRESULT WINAPI HTMLSelectElementEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
 {
     HTMLSelectElementEnum *This = impl_from_IEnumVARIANT(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
-    return E_NOTIMPL;
+    HTMLSelectElementEnum *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumVARIANT_iface.lpVtbl = This->IEnumVARIANT_iface.lpVtbl;
+    ret->ref = 1;
+    ret->iter = This->iter;
+    IHTMLSelectElement_AddRef(&This->elem->IHTMLSelectElement_iface);
+    ret->elem = This->elem;
+
+    *ppEnum = &ret->IEnumVARIANT_iface;
+    return S_OK;
 }
 
 static const IEnumVARIANTVtbl HTMLSelectElementEnumVtbl = {
@@ -857,8 +870,10 @@ static HRESULT WINAPI HTMLSelectElement_put_onchange(IHTMLSelectElement *iface,
 static HRESULT WINAPI HTMLSelectElement_get_onchange(IHTMLSelectElement *iface, VARIANT *p)
 {
     HTMLSelectElement *This = impl_from_IHTMLSelectElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_CHANGE, p);
 }
 
 static HRESULT WINAPI HTMLSelectElement_put_selectedIndex(IHTMLSelectElement *iface, LONG v)
diff --git a/dlls/mshtml/htmltable.c b/dlls/mshtml/htmltable.c
index 92dd4b5..19e5b52 100644
--- a/dlls/mshtml/htmltable.c
+++ b/dlls/mshtml/htmltable.c
@@ -2128,15 +2128,19 @@ static HRESULT WINAPI HTMLTable_get_readyState(IHTMLTable *iface, BSTR *p)
 static HRESULT WINAPI HTMLTable_put_onreadystatechange(IHTMLTable *iface, VARIANT v)
 {
     HTMLTable *This = impl_from_IHTMLTable(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_READYSTATECHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLTable_get_onreadystatechange(IHTMLTable *iface, VARIANT *p)
 {
     HTMLTable *This = impl_from_IHTMLTable(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_READYSTATECHANGE, p);
 }
 
 static const IHTMLTableVtbl HTMLTableVtbl = {
diff --git a/dlls/mshtml/htmltextarea.c b/dlls/mshtml/htmltextarea.c
index 0901489..63ea4b1 100644
--- a/dlls/mshtml/htmltextarea.c
+++ b/dlls/mshtml/htmltextarea.c
@@ -195,29 +195,37 @@ static HRESULT WINAPI HTMLTextAreaElement_select(IHTMLTextAreaElement *iface)
 static HRESULT WINAPI HTMLTextAreaElement_put_onchange(IHTMLTextAreaElement *iface, VARIANT v)
 {
     HTMLTextAreaElement *This = impl_from_IHTMLTextAreaElement(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_CHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLTextAreaElement_get_onchange(IHTMLTextAreaElement *iface, VARIANT *p)
 {
     HTMLTextAreaElement *This = impl_from_IHTMLTextAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_CHANGE, p);
 }
 
 static HRESULT WINAPI HTMLTextAreaElement_put_onselect(IHTMLTextAreaElement *iface, VARIANT v)
 {
     HTMLTextAreaElement *This = impl_from_IHTMLTextAreaElement(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_node_event(&This->element.node, EVENTID_SELECT, &v);
 }
 
 static HRESULT WINAPI HTMLTextAreaElement_get_onselect(IHTMLTextAreaElement *iface, VARIANT *p)
 {
     HTMLTextAreaElement *This = impl_from_IHTMLTextAreaElement(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_node_event(&This->element.node, EVENTID_SELECT, p);
 }
 
 static HRESULT WINAPI HTMLTextAreaElement_put_readOnly(IHTMLTextAreaElement *iface, VARIANT_BOOL v)
diff --git a/dlls/mshtml/mshtml_private.h b/dlls/mshtml/mshtml_private.h
index fd17a6e..0abc075 100644
--- a/dlls/mshtml/mshtml_private.h
+++ b/dlls/mshtml/mshtml_private.h
@@ -1268,6 +1268,8 @@ void nsAString_Finish(nsAString*);
 HRESULT map_nsresult(nsresult);
 HRESULT return_nsstr(nsresult,nsAString*,BSTR*);
 HRESULT return_nsstr_variant(nsresult,nsAString*,unsigned,VARIANT*);
+enum sg_url_part { SG_URLPART_HOST, SG_URLPART_HOSTNAME, SG_URLPART_PORT, SG_URLPART_PATHNAME, SG_URLPART_PROTOCOL, SG_URLPART_SEARCH, SG_URLPART_HASH };
+HRESULT url_with_part(const WCHAR*,enum sg_url_part,const WCHAR*,BSTR*);
 HRESULT variant_to_nsstr(VARIANT*,BOOL,nsAString*);
 HRESULT return_nsform(nsresult,nsIDOMHTMLFormElement*,IHTMLFormElement**);
 
diff --git a/dlls/mshtml/nsiface.idl b/dlls/mshtml/nsiface.idl
index 43786b4..4ee4e02 100644
--- a/dlls/mshtml/nsiface.idl
+++ b/dlls/mshtml/nsiface.idl
@@ -2093,33 +2093,35 @@ interface nsIDOMHTMLAnchorElement : nsISupports
 ]
 interface nsIDOMHTMLAreaElement : nsISupports
 {
-    nsresult GetAlt(const nsAString *aAlt);
+    nsresult GetAlt(nsAString *aAlt);
     nsresult SetAlt(nsAString *aAlt);
-    nsresult GetCoords(const nsAString *aCoords);
+    nsresult GetCoords(nsAString *aCoords);
     nsresult SetCoords(nsAString *aCoords);
-    nsresult GetShape(const nsAString *aShape);
+    nsresult GetShape(nsAString *aShape);
     nsresult SetShape(nsAString *aShape);
-    nsresult GetHref(const nsAString *aHref);
+    nsresult GetHref(nsAString *aHref);
     nsresult SetHref(nsAString *aHref);
-    nsresult GetTarget(const nsAString *aTarget);
+    nsresult GetTarget(nsAString *aTarget);
     nsresult SetTarget(nsAString *aTarget);
-    nsresult GetPing(const nsAString *aPing);
+    nsresult GetPing(nsAString *aPing);
     nsresult SetPing(nsAString *aPing);
-    nsresult GetDownload(const nsAString *aDownload);
+    nsresult GetDownload(nsAString *aDownload);
     nsresult SetDownload(nsAString *aDownload);
-    nsresult GetProtocol(const nsAString *aProtocol);
+    nsresult GetProtocol(nsAString *aProtocol);
     nsresult SetProtocol(nsAString *aProtocol);
-    nsresult GetHost(const nsAString *aHost);
+    nsresult GetHost(nsAString *aHost);
     nsresult SetHost(nsAString *aHost);
-    nsresult GetPort(const nsAString *aPort);
+    nsresult GetHostname(nsAString *aHostname);
+    nsresult SetHostname(nsAString *aHostname);
+    nsresult GetPort(nsAString *aPort);
     nsresult SetPort(nsAString *aPort);
-    nsresult GetPathname(const nsAString *aPathname);
+    nsresult GetPathname(nsAString *aPathname);
     nsresult SetPathname(nsAString *aPathname);
-    nsresult GetSearch(const nsAString *aSearch);
+    nsresult GetSearch(nsAString *aSearch);
     nsresult SetSearch(nsAString *aSearch);
-    nsresult GetHash(const nsAString *aHash);
+    nsresult GetHash(nsAString *aHash);
     nsresult SetHash(nsAString *aHash);
-    nsresult GetNoHref(const nsAString *aNoHref);
+    nsresult GetNoHref(nsAString *aNoHref);
     nsresult SetNoHref(nsAString *aNoHref);
     nsresult ToString(nsAString *_retval);
 }
